[FEAT] 로그인 성공 후 개인정보 동의

stopstone opened the pull request:
## 작업 내용
- 로그인 성공 후 바텀시트를 열어 개인정보를 동의합니다.
- 체크박스 전체가 체크되어야 버튼이 활성화 됩니다.
- 각 항목을 클릭하면 웹뷰로 전환됩니다.
- 웹뷰에서 다시 복귀했을 때 바텀시트를 유지합니다.

<table>
  <tr>
    <td align="center">
     <img width="300" alt="image" src="https://github.com/user-attachments/assets/d027f145-24da-4267-afd8-148d117001bd" /><br>
      체크 안됨
    </td>
    <td align="center">
<img width="300" height="1447" alt="image" src="https://github.com/user-attachments/assets/07f85df6-dc6f-4f54-a670-2a087f410c08" /><br>
      전체 체크 완료
    </td>
  </tr>
</table>


## 확인 방법
- feature/login 에 PrivacyConsentBottomSheet를 구성하였습니다.

## 참고 사항
- 로그인 성공 후 바텀시트를 열게 되는데 내리면 로그인도 실패처리를 해야하는지
- 바텀시트가 내려가지 않게 해야하는지 이야기해보면 좋을 것 같습니다!
- 개인정보 동의 상태를 따로 저장하거나 하지 않고 ui단에서만 처리하고 있습니다.
-  현재는 동의 후 **다음** 버튼을 클릭하면 **홈화면** 으로 이동하게 되는데, 연락처 주기 설정 화면으로 구성하는 것은 feat/friend-contact-cycle 브랜치에서 담당합니다.

## 관련 이슈
- close #30

stopstone committed:
refactor: 로그인 이벤트 통합

- 로그인 성공, 에러 이벤트를 `LoginEvent` sealed class로 통합하여 관리
- 약관동의를 위한 ShowPrivacyBottomShhet Event 추가

stopstone committed:
refactor: 로그인 이벤트 통합

- 로그인 성공, 에러 이벤트를 `LoginEvent` sealed class로 통합하여 관리
- 약관동의를 위한 ShowPrivacyBottomShhet Event 추가

stopstone committed:
feat: 약관 동의 바텀시트 UI 구성

- 사용자가 소셜 로그인 시, 약관에 동의할 수 있도록 바텀시트 UI를 구현했습니다.
- `PrivacyConsentBottomSheet`와 `TermsAgreementItem` Composable을 추가하여 약관 동의 화면을 구성했습니다.
- 로그인 성공 후 약관 동의 바텀시트가 나타나고, 동의 완료 시 홈 화면으로 이동하는 로직을 `LoginScreen`에 추가했습니다.

stopstone committed:
feat: 약관 동의 바텀시트 UI 구성

- 사용자가 소셜 로그인 시, 약관에 동의할 수 있도록 바텀시트 UI를 구현했습니다.
- `PrivacyConsentBottomSheet`와 `TermsAgreementItem` Composable을 추가하여 약관 동의 화면을 구성했습니다.
- 로그인 성공 후 약관 동의 바텀시트가 나타나고, 동의 완료 시 홈 화면으로 이동하는 로직을 `LoginScreen`에 추가했습니다.

stopstone committed:
refactor: 체크박스 클릭 시 리플 효과 제거

- `NearCheckbox` 컴포넌트의 클릭 효과에서 물결(ripple) 효과를 제거하기 위해 `clickable`을 `onNoRippleClick` 확장 함수로 변경했습니다.

stopstone committed:
refactor: 체크박스 클릭 시 리플 효과 제거

- `NearCheckbox` 컴포넌트의 클릭 효과에서 물결(ripple) 효과를 제거하기 위해 `clickable`을 `onNoRippleClick` 확장 함수로 변경했습니다.

stopstone committed:
feat: 약관 동의 체크 버튼 구현

- 약관 동의 상태를 관리하기 위한 `TermsAgreementState` 데이터 클래스와 `TermType` enum을 추가했습니다.
- `LoginViewModel`에 약관 전체 동의 및 개별 동의 토글 로직, 약관 상세 보기 이벤트 처리 로직을 구현했습니다.
- 필수 약관에 모두 동의해야 '가입' 버튼이 활성화되도록 수정했습니다.
- 약관 텍스트 영역 클릭 시 약관 상세 웹뷰로 이동하는 이벤트를 추가했습니다.
- `NearBottomSheetDragHandle` 컴포저블을 추가하여 BottomSheet의 드래그 핸들을 공통 컴포넌트로 분리했습니다.

stopstone committed:
feat: 약관 동의 체크 버튼 구현

- 약관 동의 상태를 관리하기 위한 `TermsAgreementState` 데이터 클래스와 `TermType` enum을 추가했습니다.
- `LoginViewModel`에 약관 전체 동의 및 개별 동의 토글 로직, 약관 상세 보기 이벤트 처리 로직을 구현했습니다.
- 필수 약관에 모두 동의해야 '가입' 버튼이 활성화되도록 수정했습니다.
- 약관 텍스트 영역 클릭 시 약관 상세 웹뷰로 이동하는 이벤트를 추가했습니다.
- `NearBottomSheetDragHandle` 컴포저블을 추가하여 BottomSheet의 드래그 핸들을 공통 컴포넌트로 분리했습니다.

stopstone committed:
feat: 로그인 성공 시 약관 동의 화면 표시

- 소셜 로그인 성공 후 홈 화면으로 바로 이동하던 로직을 수정하여, 약관 동의 BottomSheet가 먼저 표시되도록 변경했습니다.
- 약관 동의가 완료된 후 홈 화면으로 이동하는 `onPrivacyConsentComplete` 함수를 추가했습니다.

stopstone committed:
feat: 로그인 성공 시 약관 동의 화면 표시

- 소셜 로그인 성공 후 홈 화면으로 바로 이동하던 로직을 수정하여, 약관 동의 BottomSheet가 먼저 표시되도록 변경했습니다.
- 약관 동의가 완료된 후 홈 화면으로 이동하는 `onPrivacyConsentComplete` 함수를 추가했습니다.

stopstone committed:
refactor: 이용약관 동의 항목 UI 개선

- 우측 끝에 > 이미지를 추가합니다

stopstone committed:
feat: 약관 동의 화면 웹뷰 연동

- TermType에 titleRes, url 추가
- LoginScreen에서 약관 상세 보기 클릭 시 웹뷰로 이동하도록 수정
- PrivacyBottomSheet에서 TermType의 titleRes를 사용하도록 수정
- NearNavHost에 onNavigateToTerms 추가하여 웹뷰 이동 로직 연결

stopstone committed:
refactor: TermType 클래스 분리

- `TermType` enum 클래스를 `LoginViewModel.kt`에서 `login/model/TermType.kt` 파일로 분리했습니다.
- `LoginScreen.kt`, `LoginViewModel.kt`, `PrivacyBottomSheet.kt`에서 분리된 `TermType` 클래스를 참조하도록 수정했습니다.

stopstone committed:
refactor: 바텀 시트 드래그 핸들 색상 변경

stopstone committed:
refactor: 개인정보 동의 바텀시트 문자열 리소스 적용

- `PrivacyBottomSheet.kt`에서 하드코딩된 문자열들을 `strings.xml`에 정의된 리소스로 변경했습니다.
- `strings.xml`에 개인정보 동의 바텀시트 관련 문자열들을 추가했습니다.

stopstone committed:
fix: PrivacyBottomSheet가 웹뷰에서 돌아왔을 때 내려가는 현상 수정

- LoginViewModel에서 약관 동의 및 바텀시트 상태를 LoginState로 통합 관리하도록 변경했습니다.
- 웹뷰에서 돌아올 때 개인정보처리방침 바텀시트가 다시 보이도록 로직을 추가했습니다.
- LoginScreen에서 Lifecycle 이벤트를 감지하여 바텀시트 복원 로직을 호출하도록 수정했습니다.

stopstone committed:
refactor: 개인정보처리방침 바텀시트 약관 항목 모델 분리

- `PrivacyBottomSheet.kt`에서 사용하던 약관 항목 `Triple`을 `TermsItem` 데이터 클래스로 분리했습니다.

stopstone committed:
Merge remote-tracking branch 'origin/feat/login-privacy' into feat/login-privacy

# Conflicts:
#	Near/app/src/main/java/com/alarmy/near/presentation/feature/login/LoginScreen.kt
#	Near/app/src/main/java/com/alarmy/near/presentation/feature/login/LoginViewModel.kt
#	Near/app/src/main/java/com/alarmy/near/presentation/feature/login/PrivacyBottomSheet.kt
#	Near/app/src/main/java/com/alarmy/near/presentation/feature/login/components/NearBottomSheetDragHandle.kt
#	Near/app/src/main/java/com/alarmy/near/presentation/feature/login/components/TermsAgreementItem.kt

stopstone commented on Near/app/src/main/java/com/alarmy/near/presentation/feature/login/components/TermsAgreementItem.kt:
```diff
@@ -0,0 +1,79 @@
+    onclick: () -> Unit,
+    showDivider: Boolean,
+) {
+    Row(
+        modifier =
+            Modifier
+                .padding(vertical = 14.dp)
+                .onNoRippleClick(onclick),
+        verticalAlignment = Alignment.CenterVertically,
+        horizontalArrangement = Arrangement.SpaceBetween,
+    ) {
+        // 체크박스 클릭 시 체크 상태만 변경
+        NearCheckbox(
+            checked = isChecked,
+            onCheckedChange = onCheckedChange,
+        )
+
+        Spacer(modifier = Modifier.size(8.dp))
+
+        // 텍스트 영역 클릭 시 웹뷰로 이동
+        Text(
+            text = text,
+            style = NearTheme.typography.B2_14_MEDIUM,
+        )
+
+        Spacer(modifier = Modifier.weight(1f))
+
+        Image(
+            painter = painterResource(R.drawable.ic_front_24_gray),
+            contentDescription = null,
+        )
+    }
```
현재 Row 자체를 클릭하면 웹뷰로 전환되도록 의도하여 이부분은 수정하지 않겠습니다.
체크박스 부분은 웹뷰로 전환되지 않고 체크만 되고, 다른 Row영역은 클릭되는 것이 맞습니다.

stopstone committed:
refactor: 약관 동의 항목 생성 로직 개선

- `TermsItem` 데이터 클래스 사용을 제거하고, `TermType` enum class를 직접 사용하여 약관 항목을 구성하도록 수정했습니다.
- 약관 타입 목록(`terms`)을 `remember`를 사용하여 리컴포지션 시 재생성되지 않도록 최적화했습니다.

stopstone committed:
fix: 로그인 실패 시 에러 스낵바 표시

- `LoginRoute`에 `onShowErrorSnackBar` 콜백을 추가하여 로그인 실패 시 스낵바를 표시하도록 구현했습니다.
- `LoginEvent.ShowError` 발생 시 `onShowErrorSnackBar`를 호출하여 에러 메시지를 전달합니다.

stopstone committed:
refactor: delay 대신 콜백을 활용하여 웹뷰 이동

- `LoginScreen`에서 약관 클릭 시 웹뷰로 이동하는 `onNavigateToWebView` 콜백을 `PrivacyBottomSheet`에 전달하도록 수정했습니다.
- 기존 delay()를 활용한 이유는 바텀시트가 내려가기 전에 웹뷰로 전환되어 딜레이를 호출하였습니다.

stopstone committed:
feat: 웹뷰 로딩 중 프로그레스 바 표시

- `WebViewFrame.kt`에서 웹뷰 페이지 로딩 중 `CircularProgressIndicator`를 표시하도록 수정했습니다.
- 웹뷰 로딩 상태를 관리하기 위해 `isLoading` 상태 변수를 추가했습니다.
- `WebViewClient`의 `onPageStarted` 및 `onPageFinished` 콜백을 사용하여 로딩 상태를 업데이트합니다.

rhkrwngud445 commented on Near/app/src/main/java/com/alarmy/near/presentation/feature/login/PrivacyBottomSheet.kt:
```diff
@@ -0,0 +1,198 @@
+    onDismiss: () -> Unit,
+    onConsentComplete: () -> Unit,
+    onTermsClick: (TermType) -> Unit = {},
+    viewModel: LoginViewModel,
```
viewModel을 인자로 받아오는 것보다 상태를 호이스팅하는 것은 어떨까요? 해당 컴포저블은 preview 작성이 어려워 보이네요!

rhkrwngud445 reviewed (commented):
고생하셨습니다! viewModel 컴포저블 파라미터 관련해서 코멘트 남겼는데, 이것만 확인 부탁드려요!

stopstone committed:
refactor: `PrivacyBottomSheet`의 상태 호이스팅 형태로 수정

- `PrivacyBottomSheet.kt`에서 ViewModel을 직접 참조하지 않고, `LoginScreen.kt`를 통해 상태와 이벤트를 전달받도록 수정했습니다.
- `LoginScreen.kt`에서 `termsAgreementState`를 수집하고 `PrivacyBottomSheet`에 전달하도록 변경했습니다.

stopstone committed:
refactor: PrivacyBottomSheet 컨텐츠 분리 및 Preview 추가

- PrivacyBottomSheet의 컨텐츠를 `PrivacyConsentBottomSheetContent` Composable 함수로 분리했습니다.
- 분리된 `PrivacyConsentBottomSheetContent`에 대한 프리뷰를 추가했습니다.
- `PrivacyConsentBottomSheet` 자체를 프리뷰로 그리면 바텀시트가 표시되지 않는 현상이 있어 Content로 따로 분리하였습니다
